src/BaGet.Core/Decompiler/AssemblyDecompilerService.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection.PortableExecutable;
+using BaGet.Core.Decompiler.Decompilation;
+using BaGet.Core.Decompiler.SourceCode;
+using BaGet.Core.Entities;
+using BaGet.Core.SourceCode;
+using ICSharpCode.Decompiler;
+using ICSharpCode.Decompiler.CSharp;
+using ICSharpCode.Decompiler.CSharp.OutputVisitor;
+using ICSharpCode.Decompiler.Metadata;
+using ICSharpCode.Decompiler.Output;
+using ICSharpCode.Decompiler.TypeSystem;
+
+namespace BaGet.Core.Decompiler
+{
+    public class AssemblyDecompilerService
+    {
+        private readonly Filter _filter;
+        private readonly CSharpAmbience _ambience;
+
+        private readonly List<ISourceCodeProvider> _globalSourceCodeProviders;
+
+        public AssemblyDecompilerService()
+        {
+            _ambience = new CSharpAmbience
+            {
+                ConversionFlags = ConversionFlags.StandardConversionFlags
+            };
+
+            _filter = new Filter();
+
+            _globalSourceCodeProviders = new List<ISourceCodeProvider>();
+
+            // TODO: Make source code provider for SourceLink
+            // TODO: Make source code provider for embedded source in nupkg
+            // TODO: Make source code provider for embedded source in pdb
+        }
+
+        public SourceCodeAssembly AnalyzeAssembly(Stream assembly, Stream pdb = null, Stream documentationXml = null)
+        {
+            var localSourceCodeProviders = new List<ISourceCodeProvider>();
+
+            var assemblyPe = new PEFile("Binary.dll", assembly, PEStreamOptions.PrefetchEntireImage);
+
+            // TODO: Read PDB's (are they even needed?)
+            //PEFile pdbPe = null;
+            //if (pdb != null)
+            //    pdbPe = new PEFile("Binary.pdb", pdb, PEStreamOptions.PrefetchEntireImage);
+
+            var decompiler = new CSharpDecompiler(assemblyPe, new NullAssmblyResolver(), new DecompilerSettings(LanguageVersion.Latest));
+            var res = DecompileAssembly(decompiler.TypeSystem.MainModule);
+
+            // Read documentation
+            // TODO Read documentation
+
+            // Fetch sources
+            localSourceCodeProviders.Add(new SourceCodeDecompiler(decompiler));
```

Comment:
Why is there a distinction between global and local source code providers? Could `SourceCodeDecompiler` be part of `_globalSourceCodeProviders `?

Reply:
I initially planned this as a series of "tryDecompileSource" methods, where Globals can run for all decompilations while Locals are specific for this one run of decompilation. We try not to create too many objects where I come from :).

The whole transient-preferred mindset is new to me. In that light, every decompiler is "local" (create pr. use), so this global thing can go.